contracts/BadgerRegistry.sol

```diff
@@ -70,112 +105,196 @@ contract BadgerRegistry {
+    bool addedToVersionStatusSet = productionVaults[version][actualStatus].add(vault);
+    // If addedToVersionStatusSet remove from old and emit event
+    if (addedToVersionStatusSet) {
       // also remove from old prod
-      if(uint256(actualStatus) == 2){
-        // Remove from prev2
-        productionVaults[version][VaultStatus(0)].remove(vault);
-        productionVaults[version][VaultStatus(1)].remove(vault);
-      }
-      if(uint256(actualStatus) == 1){
-        // Remove from prev1
-        productionVaults[version][VaultStatus(0)].remove(vault);
+      if (uint256(actualStatus) > 0) {
+        for (uint256 status_ = uint256(actualStatus); status_ > 0; --status_) {
+          productionVaults[version][VaultStatus(status_ - 1)].remove(vault);
+        }
       }
 
-      emit PromoteVault(msg.sender, version, vault, actualStatus);
+      emit PromoteVault(msg.sender, version, metadata, vault, actualStatus);
     }
   }
 
-  function demote(string memory version, address vault, VaultStatus status) public {
-    require(msg.sender == governance || msg.sender == devGovernance, "!gov");
+  function demote(address vault, VaultStatus status) public {
+    require(msg.sender == governance || msg.sender == strategistGuild || msg.sender == devGovernance, "!auth");
 
     VaultStatus actualStatus = status;
-    if(msg.sender == devGovernance) {
+    if (msg.sender == devGovernance) {
       actualStatus = VaultStatus.experimental;
     }
 
-    bool removed = productionVaults[version][actualStatus].remove(vault);
+    VaultInfo memory existedVaultInfo = productionVaultInfoByVault[vault];
+    require(existedVaultInfo.vault != address(0), "BadgerRegistry: Vault does not exist");
+    // Value should be allowed to be equal to allow for promotion of vaults to experimental status as prod vaults
+    require(uint256(actualStatus) < uint256(existedVaultInfo.status), "BadgerRegistry: Vault is not being demoted");
+
+    productionVaults[existedVaultInfo.version][existedVaultInfo.status].remove(vault);
+    productionVaultInfoByVault[vault].status = status;
+    emit DemoteVault(msg.sender, existedVaultInfo.version, existedVaultInfo.metadata, vault, status);
+    productionVaults[existedVaultInfo.version][status].add(vault);
+  }
+
+  function purge(address vault) public {
+    require(msg.sender == governance || msg.sender == strategistGuild, "!auth");
 
-    if (removed) {
-      emit DemoteVault(msg.sender, version, vault, status);
+    VaultInfo memory existedVaultInfo = productionVaultInfoByVault[vault];
+    require(existedVaultInfo.vault != address(0), "BadgerRegistry: Vault does not exist");
+
+    bool removedFromVersionStatusSet = productionVaults[existedVaultInfo.version][existedVaultInfo.status].remove(
+      vault
+    );
+    bool deletedFromVaultInfoByVault = productionVaultInfoByVault[vault].vault != address(0);
+    if (removedFromVersionStatusSet || deletedFromVaultInfoByVault) {
+      delete productionVaultInfoByVault[vault];
```

Comment:
I remember a bug where deleting a storage of a struct causes only the first value to be deleted. May need to set everything back to 0 / default